src/DssDirectDepositHub.sol

```diff
@@ -0,0 +1,571 @@
+    EndLike      public           end;
+
+    /**
+        @notice Tracking struct for each of the D3M ilks.
+        @param pool   Contract to access external pool and hold balances
+        @param plan   Contract used to calculate target debt
+        @param tau    Time until you can write off the debt [sec]
+        @param culled Debt write off triggered (1 or 0)
+        @param tic    Timestamp when the pool is caged
+    */
+    struct Ilk {
+        D3MPoolLike pool;   // Access external pool and holds balances
+        D3MPlanLike plan;   // How we calculate target debt
+        uint256     tau;    // Time until you can write off the debt [sec]
+        uint256     culled; // Debt write off triggered
+        uint256     tic;    // Timestamp when the pool is caged
+    }
+
+    /// @notice maps ilk bytes32 to the D3M tracking struct.
+    mapping (bytes32 => Ilk) public ilks;
+
+    // --- Events ---
+    event Rely(address indexed usr);
+    event Deny(address indexed usr);
+    event File(bytes32 indexed ilk, bytes32 indexed what, address data);
+    event File(bytes32 indexed ilk, bytes32 indexed what, uint256 data);
+    event File(bytes32 indexed what, address data);
+    event Wind(bytes32 indexed ilk, uint256 amount);
+    event Unwind(bytes32 indexed ilk, uint256 amount, uint256 fees);
+    event Reap(bytes32 indexed ilk, uint256 amt);
+    event Cage();
+    event Cage(bytes32 indexed ilk);
+    event Cull(bytes32 indexed ilk);
+    event Uncull(bytes32 indexed ilk);
+    event Quit(bytes32 indexed ilk, address indexed usr);
+    event Exit(bytes32 indexed ilk, address indexed usr, uint256 amt);
+
+    /**
+        @dev sets msg.sender as authed.
+        @param vat_     address of the DSS vat contract
+        @param daiJoin_ address of the DSS Dai Join contract
+    */
+    constructor(address vat_, address daiJoin_) public {
+        vat = VatLike(vat_);
+        daiJoin = DaiJoinLike(daiJoin_);
+        TokenLike(DaiJoinLike(daiJoin_).dai()).approve(daiJoin_, type(uint256).max);
+        VatLike(vat_).hope(daiJoin_);
+
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+    }
+
+    // --- Math ---
+    uint256 internal constant RAY = 10 ** 27;
+
+    function _add(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require((z = x + y) >= x, "DssDirectDepositHub/overflow");
+    }
+    function _sub(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require((z = x - y) <= x, "DssDirectDepositHub/underflow");
+    }
+    function _mul(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require(y == 0 || (z = x * y) / y == x, "DssDirectDepositHub/overflow");
+    }
+    function _min(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = x <= y ? x : y;
+    }
+    function _max(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = x >= y ? x : y;
+    }
+    function _divup(uint256 x, uint256 y) internal pure returns (uint256 z) {
```

Comment:
@kmbarry1 has divup ever been formally verified or spec'ed in any previous projects that use it?

Reply:
Not AFAIK but this isn't really the sort of thing I'd be concerned about needing FV.

Reply:
switch to:
```solidity
unchecked {
    z = x != 0 ? ((x - 1) / y) + 1 : 0;
}
```

Reply:
done 4383bfe